### ISSUE TITLE
Touch controls for Android

### ISSUE THREAD

**Issue by selaux** [ja2-stracciatella/ja2-stracciatella#1250]
This ticket should be a basis to discuss how touch controls should be implemented. Here is my suggestion:

Implement touch controls as a cli option/launcher setting: `-touchcontrols`.

For most mouse regions (such as buttons):
- Short touch = Left Click
- Long touch = Right Click

For tactical map:
- Panning using a single finger touch
- Single touch moves cursor / selects merc
- Not sure how to do aiming, etc. yet

Edge cases:
- How to do things for which there is only a shortcut, e.g. running

**Comment by lynxlynxlynx** [comment]
Why a -touchcontrols? SDL fires separate touch events, so there's no complication by supporting them by default.

Maybe pinching for aiming? Seems kind of natural.

**Comment by selaux** [comment]
SDL2 also fires click events for touch events. It seems possible to ignore those though. So yes it could be possible to omit the `-touchcontrols`.

Yes pinching sounds good for aiming.

**Comment by selaux** [comment]
I'm trying to collect feedback into this ticket

From #1338:

> Still testing JA2S version 0.17.1 on Android 9 on my Galaxy S8, I found myself struggling with not having access to the inventory in the tactical screen. Main problem is that there is no mouse right-click possible, but also that there is apparently no way to open the keyboard to use the shortcut instead.

> Since the screen format is 4:3 for the game there would be plenty of screen space left to add some buttons for right-click and toggling the keyboard, what do you think?

Issue: Getting into the inventory screen in tactical mode

**Comment by jfkur** [comment]
Hi. I wanted to play JA2 on my phone and lack of proper touch controls is i think game breaker if there is no option to use keyboard and mouse. So i I've created first attempt to implement those. My work can be seen on my fork's own pull request:

https://github.com/jfkur/ja2-stracciatella/pull/1/files

I want to note that am not familiar to much with developing apps on android. In my day job i am working mostly on backends written in Java.

Going back to what i did. Core changes are made to SDLSurface class where I've added touch gestures. The rest of changes were forced by app not compiling. Current behavior is as follows:
- Single tap confirmed - Left mouse button click
- Long press - Double left mouse button click
- Double tap - Right mouse button click
- Scroll - Scrolling of a map (by pressing arrow keys underneath) - this is very cheap implementation not working smooth
- Fling - Some activities based on direction (up - stand, down - crouch, left - prone,  right - stealth)
- Scale - toggle cursor level

Maybe what i did will help with implementation of touch gestures.

BTW. pinch gesture for aiming will be very tricky to implement because of multiple fingers touching the screen and accidentally moving cursor away from target.

**Comment by selaux** [comment]
I think this might be a viable intermediate solution until we get touch controls into Stracciatella itself. From what I understand it would be some kind of compatibility layer for Android (similar to what [the previous Android](http://thepit.ja-galaxy-forum.com/index.php?t=msg&th=17937&start=0&) port had). In the short term this might be useful in order to make it playable somehow, but I think in the long term the inability to inspect the game state will limit this approach (e.g. fixing the fat finger problem you mentioned). I'm open to try it out though.

The approach that I had in mind was creating a new PointerSystem, similar to the existing [MouseSystem](https://github.com/ja2-stracciatella/ja2-stracciatella/blob/master/src/sgp/MouseSystem.cc), to be able to handle touch and mouse events (and make the MouseSystem a compatibility layer around it). Then we could start using the touch-specific events from the PointerSystem where we think it is useful. SDL2 itself has capabilities around touch controls, so it should be possible. The advantages of this approach would be that also non-android systems could make use of these touch controls and they can be more easily integrated with the game itself (e.g. we could show touch specific controls and such). I started a few times with it but it's quite some effort, that is why I think this intermediate solution could be viable.

Long story short I will have a look at what you implemented until now and check how well it works in practice and get back to you as soon as possible.

**Comment by jfkur** [comment]
@selaux Yes i can agree that what i did is some kind of workaround to what should be really implemented. Unfortunately i do not have experience and knowledge in SDL2 and its capabilities :) 
@lynxlynxlynx Maybe something like this could work but it would need more changes than only handling gestures itself

BTW. I've removed scroll behavior as it seems that in this state it is not working well. It seems that moving finger a little is recognized as a scroll and it prevents moving cursor around when arrow key was pressed. In this state it was hard to place cursor on desired target

**Comment by lynxlynxlynx** [comment]
Another option instead of pinch would be relying on physical buttons. Is there android hw without +/- volume buttons? I'd bet the majority have them.

**Comment by lynxlynxlynx** [comment]
Also an opportunity to remove WITH_MAEMO.

**Comment by selaux** [comment]
I think we can close this now. Basic touch controls are there. They only use a few gestures (long touch for right click, double tap for double click, two fingers for scrolling in tactical). Actions in tactical have their own dedicated controls. While in the beginning I though that gestures would be nice for that, I think adding a lot of gestures just makes the controls harder to learn and requires more documentation. Let's open more specific issues for issues with touch controls now.
